mapeditor/campaigneditor/startingbonus.cpp

```diff
@@ -0,0 +1,370 @@
+		bonus.type = CampaignBonusType::SECONDARY_SKILL;
+	else if(ui->radioButtonResource->isChecked())
+		bonus.type = CampaignBonusType::RESOURCE;
+
+	bonus.info1 = 0;
+	bonus.info2 = 0;
+	bonus.info3 = 0;
+
+	switch (bonus.type)
+	{
+	case CampaignBonusType::SPELL:
+		bonus.info1 = ui->comboBoxSpellRecipient->currentData().toInt();
+		bonus.info2 = ui->comboBoxSpellSpell->currentData().toInt();
+		break;
+	case CampaignBonusType::MONSTER:
+		bonus.info1 = ui->comboBoxCreatureRecipient->currentData().toInt();
+		bonus.info2 = ui->comboBoxCreatureCreatureType->currentData().toInt();
+		bonus.info3 = ui->spinBoxCreatureQuantity->value();
+		break;
+	case CampaignBonusType::BUILDING:
+		bonus.info1 = ui->comboBoxBuildingBuilding->currentData().toInt();
+		break;
+	case CampaignBonusType::ARTIFACT:
+		bonus.info1 = ui->comboBoxArtifactRecipient->currentData().toInt();
+		bonus.info2 = ui->comboBoxArtifactArtifact->currentData().toInt();
+		break;
+	case CampaignBonusType::SPELL_SCROLL:
+		bonus.info1 = ui->comboBoxSpellScrollRecipient->currentData().toInt();
+		bonus.info2 = ui->comboBoxSpellScrollSpell->currentData().toInt();
+		break;
+	case CampaignBonusType::PRIMARY_SKILL:
+		bonus.info1 = ui->comboBoxPrimarySkillRecipient->currentData().toInt();
+		bonus.info2 |= ui->spinBoxPrimarySkillAttack->value() << 0;
+		bonus.info2 |= ui->spinBoxPrimarySkillDefense->value() << 8;
+		bonus.info2 |= ui->spinBoxPrimarySkillSpell->value() << 16;
+		bonus.info2 |= ui->spinBoxPrimarySkillKnowledge->value() << 24;
+		break;
+	case CampaignBonusType::SECONDARY_SKILL:
+		bonus.info1 = ui->comboBoxSecondarySkillRecipient->currentData().toInt();
+		bonus.info2 = ui->comboBoxSecondarySkillSecondarySkill->currentData().toInt();
+		bonus.info3 = ui->comboBoxSecondarySkillMastery->currentData().toInt();
+		break;
+	case CampaignBonusType::RESOURCE:
+		bonus.info1 = ui->comboBoxResourceResourceType->currentData().toInt();
+		bonus.info2 = ui->spinBoxResourceQuantity->value();
+		break;
+	
+	default:
+		break;
+	}
+}
+
+void StartingBonus::on_buttonBox_clicked(QAbstractButton * button)
+{
+	if(button == ui->buttonBox->button(QDialogButtonBox::Ok))
+	{
+		saveBonus();
+		accept();
+	}
+	close();
+}
+
+bool StartingBonus::showStartingBonus(PlayerColor color, std::shared_ptr<CMap> map, CampaignBonus & bonus)
+{
+	auto * dialog = new StartingBonus(color, map, bonus);
+
+	dialog->setAttribute(Qt::WA_DeleteOnClose);
+
+	auto result = dialog->exec();
+
+	if(result == QDialog::Accepted)
+		bonus = dialog->bonus;
+
+	return result == QDialog::Accepted;
+}
+
+QString StartingBonus::getBonusListTitle(CampaignBonus bonus, std::shared_ptr<CMap> map)
+{
+	auto getHeroName = [](int id){
+		MetaString tmp;
+		if(id == 0xFFFD)
+			tmp.appendRawString(tr("strongest hero").toStdString());
+		else if(id == 0xFFFE)
+			tmp.appendRawString(tr("generated hero").toStdString());
+		else if(id == 0xFFFF)
+			tmp.appendRawString(tr("random hero").toStdString());
```

Comment:
These values should be be constants in HeroTypeID class, similar to HeroTypeID::NONE, if they are not already there.